design/design-doc-aggregate-resources.md

```diff
@@ -0,0 +1,965 @@
+   `ConfigureGKECluster` configurator could be reused for this.
+1. Process any `GKECluster` managed resources bound to the `AggregateResource`,
+   updating their binding status and propagating their connection secrets.
+
+No controller is watching for `AggregateResource` specifically; only resource
+claims that use or would use (due to their class) an `AggregateResource`. This
+means that a statically provisioned `AggregateResource` will not bind to any of
+the managed resources it would aggregate until it is claimed.
+
+This design implies the introduction of resource classes for managed resources
+that do not currently support them; there is no `ResourceGroupClass`, for
+example, because `ResourceGroup` has never had a corresponding resource claim to
+bind to. Furthermore, each infrastructure stack must instantiate an
+aggregation-aware resource claim controller for each possible (resource claim,
+managed resource) tuple. Put otherwise, if it makes _any_ sense for a particular
+managed resource to aggregate to a particular claim kind a controller must own
+that relationship. Many supporting managed resources map only to a single
+resource claim kind. A `MySQLServerVirtualNetworkRuleClass` is unlikely to be
+useful in the context of a `RedisCluster` resource claim; it likely need only
+apply to `MySQLServer` claims. Several resources however, including
+`ResourceGroup` and networking constructs like `Subnetwork` could make sense to
+aggregate to almost all resource claim kinds.
+
+Care must be taken when naming a dynamically provisioned `AggregateResource`,
+because creation of the `AggregateResource` could be handled any aggregate-aware
+resource claim controller concerned with one of the aggregated resource classes.
+Dynamic provisioning [typically uses] the API server's built in `GenerateName`
+support to generate a non-deterministic name for the newly created managed
+resource. If two controllers were to use this approach there would be a race;
+each controller would create a differently named `AggregateResource`, but only
+one would succeed in updating the `resourceRef` of the resource claim, leaving
+an orphaned `AggregateResource`. This can be avoided by emulating `GenerateName`
+locally, and using a name-then-create approach:
+
+1. If the resource claim has a `resourceRef.name`, but no `resourceRef.uid`, all
+   concerned controllers race to create the named `AggregateResource`. The
+   create will fail if another controller created it first. If the create
+   succeeds, the controller writes the newly created `AggregateResource` UID to
+   `.resourceRef.uid`.
+1. If the claim has no `resourceRef.name` all concerned controllers generate a
+   name using the typical managed resource naming scheme and attempt to set it
+   as the claim's `resourceRef.name`. The write will fail if another controller
+   has set the name since the controller read the claim.
+
+## Alternatives Considered
```

Comment:
I have a number of concerns above, but very much agree that this design is superior to all listed below. Thank you for outlining the alternatives as it makes your chosen design much easier to reason about :)